README.md
web-camera
==========

网页截图工具 (by phantomjs)。通过phantomjs来打开渲染网页，对网页进行截图。

## Usage

```js
var Camera = require('webcamera');
var fs = require('fs');

/**
 * web camera by Node.js and Phantomjs
 * @param {Object} options
 *   - path          {String}   default picture dir path
 *   - workerNum     {Number}   child_process max num
 *   - timeout       {Number}   child_process timeout.
 *   - phantom       {String}   phantomjs path
 *   - phantomScript {String}   phantomjs script path, use input arguments as default script
 *   - qnClient      {Object}   qiniu client instance
 *   - qnOpts        {Object}   qiniu options. if do not have qnClient and qnOpts, shotQN become invalid
 */
var camera = Camera.create({
  qnOpts: {
    accessKey: 'accessKey',
    secretKey: 'secretKey',
    bucket: 'webcamera'
  }
});

// 当处理速度比调用速度低时会触发此事件
camera.on('overload', function (listLength) {
  //listLength为排队等待处理的长度
});

// 截图保存到本地
camera.shot('http://www.baidu.com', './baidu.png', function (err, data) {
  // data.should.equal('./baidu.png');
});

// 截图作为stream
camera.shotStream('http://www.baidu.com', function (err, s) {
  var datas = [];
  var filePath = './test.jpg';
  var file = fs.createWriteStream(filePath, {encoding: 'binary'});
  s.on('data', function (data) {
    file.write(data.toString('binary'), 'binary');
  });
  s.on('end', function () {
    console.log('get pictrue ok');
  });
});

//截图上传到七牛空间，第二个参数为上传七牛的options，第三个参数是截图的options
camera.shotQN('http://www.baidu.com', {key: 'test/baidu.png'}, {quality: 10}, function (err, data) {
  /*
  data.should.like:
  ({
    hash: 'FlDGti9pVGQ3sw2oao-mVu3nZWjZ',
    key: 'test/baidu.png',
    url: 'http://webcamera.u.qiniudn.com/test/baidu.png'
  })
  */
});
```

所有的调用都可以在`callback`之前传入参数`options`.

```js
camera.shotQN('http://www.baidu.com',320, 'baidu.png', {
  clipRect: {
    top: 0,
    left:0,
    height: 'all',
    width: 'all'
  }
}, function (err, data) {
  /*
  data.should.like:
  {name: 'L1/1/320/baidu.png', size: 36889, url: 'xxx/L1/1/320/baidu.png'}
  */
});
```

|名字|类型|含义|
|----|----|----|
|clipRect|Object|指定截图的矩形区域。有四个属性:top(0), left(0), height(window), width(window)。height和width可以设置为window或者all,window将会截取当前一屏，all会截取网页全部大小|
|viewportSize|Object|设置网页的分辨率，有两个属性:width(1024), height(768)。|
|renderDelay|Number|网页加载完成之后延迟多少毫秒之后截图，默认为0|
|picPath|String|设置图片保存位置，只在`shot`方法时生效，等效于shot方法的第二个参数|
|mimeType|String|设置截图的保存类型（只有在没设置图片保存路径的情况下生效，否则使用图片保存路径的后缀类型），支持png, jpeg, gif.默认为png|
|script|Function|网页加载完成之后可以在网页中执行这个方法。|
|quality|Number|0~100，指定生成图片的质量，数值越高质量越好|

## Install

```bash
$ npm install webcamera
```

## Dependences
* [`phantomjs`](http://phantomjs.org/) >= v1.9

## Debug

```bash
$ tail -f /tmp/phantom_shot.log &
$ phantomjs "phantom/web_camera_phantom.js" "https://github.com/" > github.png
```

github page screen shot: [github.png](http://nfs.nodeblog.org/b/0/b06ed6be50682731bfae32d79b25894b.png)

## command line tool
`npm install webcamera -g`, 之后可以使用web camera 提供的命令行工具进行网页截图。需要安装`phantomjs`。

```
camera -u http://www.google.com -o google.png

Options:
  -u, --url     Web's url
  -o, --out     Output screenshot picture path                                                  [default: "./out.png"]
  -c, --config  config file path, you can define much more options by a js file or a json file
```

可以通过传递配置文件的方式进行批量截图，并对截图的详细参数进行设置，配置文件可以是json文件或者js文件，配置文件模版：

```
#json形式
[{
  "url": "http://cnodejs.org",
  "out": "cnodejs.png"
}, {
  "url": "http://nodejs.org",
  "out": "nodejs.png",
  "options": {
    "viewportSize": {
      "width": 1280,
      "height": 800
    }
  }
}, {
  "url": "http://google.com",
  "out": "google.png",
  "options": {
    "quality": 10
  }
}]

#js形式
var url = 'http://google.com';

var config = [];

for (var quality = 10; quality <= 100; quality += 10) {
  config.push({
    url: url,
    out: 'google@' + quality + '.jpg',
    options: {
      quality: quality
    }
  });
}

module.exports = config;

```

推荐通过js形式的配置文件，将会更加灵活。

## Licences
(The MIT License)

Copyright (c) 2013 dead-horse and other contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.2.3 / 2016-10-28
==================

  * fix: listen page error event

0.2.1 / 2014-08-06
==================

  * remove link
  * add travis-ci

0.2.0 / 2014-07-10
==================

  * remove tfs

0.1.1 / 2013-11-21
==================

  * delay to eval script
  * add doc
  * support qn

0.1.0 / 2013-11-15
==================

  * update readme
  * add bin
  * add tips for taobao
  * remove repeat code
  * Release 0.0.9
  * Merge pull request #5 from fengmk2/improve-performance
  * add quality options for shot()
  * add error name
  * delete base64 size check; support open url fail detach; custom log filepath.
  * Release 0.0.8
  * Merge pull request #4 from fengmk2/add-pid-to-stream
  * test err first
  * add phantomjs pid to stream.pid
  * Release 0.0.7
  * Merge pull request #3 from fengmk2/hotfix-linux-stdout
  * hot fix linux stdout
  * add history
  * Release 0.0.6
  * Merge pull request #2 from fengmk2/render-to-stdout
  * fix debug doc
  * send to stdout directly; add logger for phantomjs script

0.0.9 / 2013-09-13
==================

  * Merge pull request #5 from fengmk2/improve-performance
  * add quality options for shot()
  * add error name
  * delete base64 size check; support open url fail detach; custom log filepath.

0.0.8 / 2013-09-12
==================

  * Merge pull request #4 from fengmk2/add-pid-to-stream
  * test err first
  * add phantomjs pid to stream.pid

0.0.7 / 2013-09-12
==================

  * Merge pull request #3 from fengmk2/hotfix-linux-stdout
  * hot fix linux stdout
  * add history

0.0.6 / 2013-09-11
==================

  * Merge pull request #2 from fengmk2/render-to-stdout
  * fix debug doc
  * send to stdout directly; add logger for phantomjs script
  * Release v0.0.5
  * Merge pull request #1 from fengmk2/patch-1
  * Update tfs and utility.
  * Release v0.0.4
  * use default if not exist
  * use blanket
  * fix readme
  * add dependese version
  * Release v0.0.3
  * Release v0.0.2
  * rm close


test/web_camera.test.js
/*!
 * camera - test/web_camera.js
 * Author: dead_horse <dead_horse@qq.com>
 */

'use strict';

/**
 * Module dependencies.
 */
var Camera = require('../');
var mm = require('mm');
var child_process = require('child_process');
var should = require('should');
var fs = require('fs');
var pedding = require('pedding');

var qnOpts = {
  accessKey: 'test',
  secretKey: 'haha',
  bucket: 'test'
};

var camera;
var noTFSCamera;

describe('lib/web_camera.js', function () {
  afterEach(mm.restore);

  describe('#create', function () {
    it('should create use default', function () {
      noTFSCamera = Camera.create();
      should.not.exist(noTFSCamera.tfsClient);
    })

    it('should create use qnOpts', function () {
       camera = Camera.create({qnOpts: qnOpts});
    })

    it('should create with wrong phantom path', function () {
      noTFSCamera = Camera.create({
        phantom: '/not/exist'
      });
      camera.phantom.should.equal('phantomjs');
    });
  });

  describe('#shot', function () {
    it('should shot default ok', function (done) {
      camera.shot(__filename, function (err, data, pid) {
        fs.existsSync(data).should.be.ok;
        fs.unlinkSync(data);
        pid.should.be.number;
        pid.should.above(0);
        done(err);
      });
    });

    it('should shot to path ok', function (done) {
      camera.shot(__filename, './baidu.png', function (err, data) {
        data.should.include('baidu.png');
        fs.existsSync(data).should.be.ok;
        fs.unlinkSync(data);
        done(err);
      });
    });

    it('should shot with options ok', function (done ) {
      camera.shot(__filename, {
        quality: 80,
        picPath: './baidu.jpg',
        clipRect: {
          top: 100,
          left: 100,
          width: 100,
          height: 100
        },
        renderDelay: 100
      }, function (err, data) {
        data.should.include('baidu.jpg');
        fs.existsSync(data).should.be.ok;
        fs.statSync(data).size.should.above(0);
        fs.unlinkSync(data);
        done(err);
      });
    });

    it('should shot with path and options ok', function (done) {
      camera.shot(__filename, 'baidu.gif', {}, function (err, data) {
        data.should.include('baidu.gif');
        fs.existsSync(data).should.be.ok;
        fs.unlinkSync(data);
        done(err);
      });
    });

    it('should shot error of phantom', function (done) {
      camera.shot('www.zcxvk213123213.com', function (err, data) {
        should.not.exist(data);
        should.exist(err);
        err.message.should.equal('phantomjs exit with code 100, open url fail');
        Array.isArray(err.args).should.be.ok;
        done();
      });
    });

    it('should return when page has error', function(done) {
      camera = Camera.create();
      camera.shot('https://os.alipayobjects.com/rmsportal/UKYoZFxFsmwxVjB.html', './test.png', function (err, data) {
        data.should.include('test.png');
        fs.existsSync(data).should.be.ok;
        fs.unlinkSync(data);
        done(err);
      });
    });
  });

  describe('#shotStream', function () {
    it('should shotStream default ok', function (done) {
      camera.shotStream(__filename, function (err, s, pid) {
        var datas = [];
        var filePath = './test.png';
        s.should.have.property('pid').with.be.Number;
        var file = fs.createWriteStream(filePath, {encoding: 'binary'});
        s.on('data', function (data) {
          file.write(data.toString('binary'), 'binary');
        });
        s.on('end', function () {
          fs.existsSync(filePath).should.be.ok;
          fs.unlinkSync(filePath);
          done();
        });
        pid.should.be.Number;
        pid.should.above(0);
      });
    });

    it('should shotStream with options ok', function (done) {
      camera.shotStream(__filename,
      {mimeType: 'jpg', clipRect: {top: 100, left: 100, width: 100, height: 100}}, function (err, s, pid) {
        var datas = [];
        var filePath = './test.jpg';
        var file = fs.createWriteStream(filePath, {encoding: 'binary'});
        s.on('data', function (data) {
          file.write(data.toString('binary'), 'binary');
        });
        s.on('end', function () {
          fs.existsSync(filePath).should.be.ok;
          fs.unlinkSync(filePath);
          done();
        });
        pid.should.be.Number;
        pid.should.with.above(0);
      });
    });
  });


  describe('overload event', function () {
    it('should emit overload', function (done) {
      mm(camera, 'workerNum', 1);
      done = pedding(3, done);
      camera.once('overload', function (num) {
        num.should.equal(1);
        done();
      });
      camera.shotStream(__filename, function (err, s) {
        should.not.exist(err);
        done();
      });
      camera.shotStream(__filename, function (err, s) {
        should.not.exist(err);
        done();
      });
    });
  });
});


phantom/web_camera_phantom.js
/*!
 * web-camera - lib/web_camera_phantom.js
 *
 * Usage:
 *
 * $ phantomjs web_camera_phantom.js 1:address 2:top 3:left 4:width 5:height 6:ViewWidth 7:viewHeight \
 *   8:output 9:quality 10:waitTime 11:timeout 12:streamType 13:userScript 14:logfile
 * 
 * Authors: 
 *   dead_horse <dead_horse@qq.com>
 *   苏千 <suqian.yf@taobao.com>
 */

var system = require('system');
var webPage = require('webpage');
var fs = require('fs');
var args = Array.prototype.slice.call(system.args).filter(function (a) {
  return !a || a.indexOf('--') !== 0;
});

var page = webPage.create();

var address = args[1];

var clipRect = {
  top: parseInt(args[2], 10) || 0,
  left: parseInt(args[3], 10) || 0,
  width: parseInt(args[4], 10) || 'all',
  height: parseInt(args[5], 10) || 'all'
};

var viewportSize = {
  width: parseInt(args[6], 10) || 'all',
  height: parseInt(args[7], 10) || 'all'
};

if (viewportSize.width !== 'all' && viewportSize.height !== 'all') {
  page.viewportSize = viewportSize;
}

var output = args[8] || '/dev/stdout';
var quality = parseInt(args[9], 10) || null;
var waitTime = parseInt(args[10], 10) || 0;
var timeout = parseInt(args[11], 10) || 120000;
var streamType = args[12] || 'png';

if (output !== '/dev/stdout') {
  streamType = output.substring(output.lastIndexOf('.') + 1);
}

var userScript = args[13] || null;
var logfile = args[14] || '/tmp/phantom_shot.log';
if (logfile.indexOf('--') === 0) {
  logfile = '/tmp/phantom_shot.log';
}

var logger = fs.open(logfile, 'a');

var start = Date.now();

var log = function () {
  var args = Array.prototype.slice.call(arguments);
  logger.writeLine('[' + system.pid + '] [' + Date() + '] [' + (Date.now() - start) + 'ms] ' + args.map(function (a) {
    return String(a);
  }).join(' '));
  logger.flush();
};

var version = phantom.version.major + '.' + phantom.version.minor + '.' + phantom.version.patch;

log('phantomjs ' + version, ', os architecture: ' + system.os.architecture, 
  ', name: ' + system.os.name,
  ', version: ' + system.os.version);

log.apply(null, Array.prototype.slice.call(system.args).map(function (a, index) {
  return '"' + a + '"';
}));

log('clipRect: ' + JSON.stringify(clipRect),
  ', viewportSize: ' + JSON.stringify(page.viewportSize),
  ', output: ' + output,
  ', waitTime: ' + waitTime,
  ', timeout: ' + timeout,
  ', streamType: ' + streamType,
  ', quality: ' + (quality ? quality : 'default'),
  ', userScript: ' + userScript,
  ', logfile: ' + logfile
);

var pageDimensions = page.evaluate(function() {
  return {
    width: Math.max( 
      document.body.offsetWidth, 
      document.body.scrollWidth, 
      document.documentElement.clientWidth, 
      document.documentElement.scrollWidth, 
      document.documentElement.offsetWidth
    ), 
    height: Math.max(
      document.body.offsetHeight,
      document.body.scrollHeight,
      document.documentElement.clientHeight,
      document.documentElement.scrollHeight,
      document.documentElement.offsetHeight
    )
  };
});

/*
 * Given a shotSize dimension, return the actual number of pixels in the 
 * dimension that phantom should render.
 *
 * @param (String) dimension
 * @param (String or Number) value
 */
var pixelCount = function(dimension, value) {
  return {
    window: args[{
      width: 'windowWidth', 
      height: 'windowHeight'
    }[dimension]],
    all: pageDimensions[dimension]
  }[value] || value;
};

// Set the rectangle of the page to render
if (clipRect.width !== 'all' && clipRect.height !== 'all') {
  page.clipRect = {
    top: clipRect.top,
    left: clipRect.left, 
    width: pixelCount('width', clipRect.width), 
    height: pixelCount('height', clipRect.height)
  };
}

var resourcesNum = 0;

var _done = false;

function finish(reason) {
  if (_done) {
    return;
  }

  if (!output || output !== '/dev/stdout') {
    if (quality) {
      page.render(output, {quality: quality});
    } else {
      page.render(output);
    }
  } else {
    var base64 = page.renderBase64(streamType);
    console.log(base64);
    log('base64 ' + streamType + ' file size: ' + base64.length);
  }
  log(reason, ', render "' + address + '" to ' + output);
  _done = true;
  phantom.exit(0);
}

var resources = {};

page.onResourceRequested = function (req, networkRequest) {
  resources[req.id] = {start: Date.now()};
  resourcesNum++;
  log('req(#' + req.id + '):', req.method, req.url, ', resources left: ' + resourcesNum);
};

page.onResourceReceived = function (res) {
  var reqInfo = resources[res.id];
  if (res.stage === 'start') {
    reqInfo.size = res.bodySize;
    return;
  }

  resourcesNum--;
  var bodySize = reqInfo.size;
  var use = Date.now() - reqInfo.start;
  // var size = page.renderBase64(streamType).length;
  log('res(#' + res.id + '):',
    res.status, 
    use + 'ms, size: ' + bodySize, res.url, 
    ', content size: ' + page.content.length, 
    // ', render base64 size: ' + size, 
    ', resources left: ' + resourcesNum);
  if (res.url === address) {
    log('page title:', page.title || 'no-title');
  }
};

// timeout之后不管怎么样都超时退出
setTimeout(function () {
  finish(timeout + 'ms timeout to finish');
}, timeout);

// 当页面未加载完，open 的回调一直未执行导致卡死，绑定 onError 可以解决这个问题
page.onError = function(msg) {
  log('onError ' + msg);
};

page.open(address, function (status) {
  log('open status: ' + status);
  if (status !== 'success' && !_done) {
    // code 100 meaning open url fail
    phantom.exit(100);
    return ;
  }
  if (waitTime) {
    setTimeout(function () {
      userScript && page.evaluate(eval('('+userScript+')'));
      finish('open "' + status + '", wait for ' + waitTime + 'ms to finish');
    }, waitTime);
  } else {
    userScript && page.evaluate(eval('('+userScript+')'));
    finish('open "' + status + '" to finish');
  }  
});


lib/web_camera.js
/*!
 * camera - lib/web_camera.js
 * Authors:
 *   dead_horse <dead_horse@qq.com>
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com/)
 */

'use strict';

/**
 * Module dependencies.
 */

var child_process = require('child_process');
var path = require('path');
var fs = require('fs');
fs.existsSync = fs.existsSync || path.existsSync;
var stream = require('stream');
var EventEmitter = require('events').EventEmitter;
var utility = require('utility');
var urlModule = require('url');
var util = require('util');
var qn = require('qn');
var tfs;
try {
  tfs = require('tfs');
} catch (err) {}

/**
 * web camera by Node.js and Phantomjs
 * @param {Object} options
 *   - path          {String}   default picture dir path
 *   - workerNum     {Number}   child_process max num
 *   - timeout       {Number}   child_process timeout.
 *   - phantom       {String}   phantomjs path
 *   - phantomScript {String}   phantomjs script path, use input arguments as default script
 *   - [phantomLog]  {String}   phantomjs log file path, default is '/tmp/phantom_shot.log'
 *   - tfsClient     {Object}   tfs client instance
 *   - tfsOpts       {Object}   tfs options. if do not have tfsClient and tfsOpts, shotTFS become invalid
 *   - qnClient      {Object}   qiniu client instance
 *   - qnOpts        {Object}   qiniu options. if do not have qnClient and qnOpts, shotQN become invalid
 */
var Camera = function (options) {
  options = options || {};
  this.path = options.path || path.join(__dirname, '../pics');
  this.workerNum = options.workerNum || 5;
  this.phantom = options.phantom || 'phantomjs';

  if (this.phantom !== 'phantomjs' && !fs.existsSync(this.phantom)) {
    console.warn(this.phantom + ' not exist, use default phantomjs');
    this.phantom = 'phantomjs';
  }

  this.phantomScript = options.phantomScript || path.join(__dirname, '../phantom/web_camera_phantom.js');
  this.phantomLog = options.phantomLog || '/tmp/phantom_shot.log';
  this.tfsClient = options.tfsClient;
  this.timeout = options.timeout;
  if (!this.tfsClient && options.tfsOpts && tfs) {
    this.tfsClient = tfs.createClient(options.tfsOpts);
  }
  this.qnClient = options.qnClient;
  if (!this.qnClient && options.qnOpts) {
    this.qnClient = qn.create(options.qnOpts);
  }
  EventEmitter.call(this);
};
util.inherits(Camera, EventEmitter);

/**
 * process arguments, generate phantomjs script's input arguments
 * @param {String} url     web page url
 * @param {Object} options input options
 * @return {Array}          phantomjs arguments
 */
Camera.prototype.processArgs = function (url, options) {
  var mimeType = options.mimeType || 'png';
  var renderDelay = options.renderDelay || 200;
  var clipRect = options.clipRect || {};
  var viewportSize = options.viewportSize || {};
  if (url[0] !== '/') {
    url = urlModule.parse(url).protocol ? url : 'http://' + url;
  }

  switch (options._type) {
  case 'tfs':
    options.mimeType = mimeType;
    options.picPath = '';
    break;
  case 'pic':
    options.picPath = options.picPath || path.join(this.path,
      utility.md5(url + JSON.stringify(clipRect) + JSON.stringify(viewportSize))) + '.' + mimeType;
    options.mimeType = '';
    break;
  case 'stream':
    options.picPath = '';
    options.mimeType = options.mimeType || 'PNG';
    break;
  case 'qn':
    options.mimeType = mimeType;
    options.picPath = '';
  }
  return [
    this.phantomScript,
    url,
    clipRect.top || 0,
    clipRect.left || 0,
    clipRect.width || 'window',
    clipRect.height || 'window',
    viewportSize.width || 1024,
    viewportSize.height || 768,
    options.picPath || '',
    options.quality || 'default',
    options.renderDelay || 0,
    this.timeout || 10000,
    options.mimeType || '',
    options.script ? String(options.script) : '',
    this.phantomLog,
    '--disk-cache=true'
  ];
};


var tasks = []; //任务队列
var notify = new EventEmitter();
var workerNum = 0;

/**
 * handle phantomjs process
 */
Camera.prototype.phantomProcess = function (args, callback) {
  workerNum++;
  var _this = this;
  var phantom = child_process.spawn(this.phantom, args);
  function finish() {
    workerNum--;
    tasks.length && _this.phantomProcess.apply(_this, tasks.shift());
  }
  var output = args[8];
  if (output && output !== '/dev/stdout') {
    // has output file
    phantom.on('exit', function (code) {
      finish();
      if (code) {
        var err = new Error('phantomjs exit with code ' + code);
        if (code === 100) {
          err.name = 'WebCameraOpenURLError';
          err.message += ', open url fail';
        }
        err.args = args;
        return callback(err);
      }
      callback(null, output, phantom.pid);
    });
  } else {
    var s = new stream.Stream();
    s.readable = true;
    s.pid = phantom.pid;
    phantom.stdout.on('data', function (data) {
      s.emit('data', new Buffer(data.toString(), 'base64'));
    });

    phantom.on('exit', function (code) {
      if (code) {
        var err = new Error('phantomjs exit with code ' + code);
        if (code === 100) {
          err.name = 'WebCameraOpenURLError';
          err.message += ', open url fail';
        }
        err.args = args;
        return s.emit('error', err);
      }
      s.emit('end');
      finish();
    });
    callback(null, s, phantom.pid);
  }
};

Camera.prototype._handleShot = function (url, options, callback) {
  var args = this.processArgs(url, options);
  tasks.push([args, callback]);
  if (workerNum < this.workerNum) {
    this.phantomProcess.apply(this, tasks.shift());
  } else {
    this.emit('overload', tasks.length);
  }
};

/**
 * shot web page, genarate a picture.
 * @param {String} url        web page url
 * @param {String} path       save path. optional
 * @param {Object} options    shot options
 *   - [quality] {Number}     图片质量, 1 - 100, 越大质量越高
 *   - clipRect {Object}      截取网页的矩形区域
 *     - top
 *     - left
 *     - width                可以为'window'或者'all',此时可以截取全屏或者全网页
 *     - height               可以为'window'或者'all',此时可以截取全屏或者全网页
 *   - viewportSize {Object}  渲染网页时的分辨率
 *     - width                default 1024
 *     - height               default 768
 *   - renderDelay  {Number}  网页加载完成之后延迟多少毫秒之后截图，默认为0
 *   - picPath      {String}  设置图片保存位置，等效于shot方法的第二个参数，(如果有第二个参数path, 会被第path参数覆盖)
 *   - mimeType     {String}  截图的格式， 没有传递保存路径的时候，会根据url, viewportSize, clipRect，mimeType生成一个名字，保存到默认的图片文件夹
 *   - script       {Function}网页加载完成之后可以在网页中执行这个方法。
 * @param {Function} callback(err, data, pid)
 *   - data {String} pictrue path
 */
Camera.prototype.shot = function (url, path, options, callback) {
  switch (arguments.length) {
  case 2: // url, callback
    callback = path;
    options = {};
    break;
  case 3: //url, path, callback | url, options, callback
    callback = options;
    if (typeof path === 'string') {
      options = {picPath: path};
    } else {
      options = path || {};
    }
    break;
  case 4: // url, path, options, callback
    options = options || {};
    options.picPath = path;
    break;
  }
  options._type = 'pic';
  this._handleShot(url, options, callback);
};

/**
 * shot web page, genarate a picture and upload to TFS.
 * @param {String} url        web page url
 * @param {Number} partition        保存到TFS时需要传入partition确定命名空间
 * @param {String} filename       保存到TFS的名字
 * @param {Object} options    shot options
 *   - clipRect {Object}      截取网页的矩形区域
 *     - top
 *     - left
 *     - width                可以为'window'或者'all',此时可以截取全屏或者全网页
 *     - height               可以为'window'或者'all',此时可以截取全屏或者全网页
 *   - viewportSize {Object}  渲染网页时的分辨率
 *     - width                default 1024
 *     - height               default 768
 *   - renderDelay  {Number}  网页加载完成之后延迟多少毫秒之后截图，默认为0
 *   - mimeType     {String}  生成截图的格式
 *   - script       {Function}网页加载完成之后可以在网页中执行这个方法。
 * @param {Function} callback(err, data)
 *   - data {String} pictrue path
 */
Camera.prototype.shotTFS = function (url, partition, filename, options, callback) {
  if (typeof filename !== 'string') {
    callback = options;
    options = filename;
    filename = 'webcamera_' + Date.now() + '_' + utility.md5(url).substring(0, 10) + '.png';
  }
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }

  if (!this.tfsClient) {
    return callback(new Error('TFS not inited'));
  }

  options = options || {};
  options._type = 'tfs';
  var self = this;
  self._handleShot(url, options, function (err, s) {
    if (err) {
      return callback(err);
    }
    self.tfsClient.uploadFile(s, partition % 1000 + 1, filename, callback);
  });
};

/**
 * shot web page, genarate a picture and upload to qiniu.
 * @param {String} url        web page url
 * @param {Object} qnOptions  上传七牛的参数，详见 https://github.com/fengmk2/qn
 * @param {Object} options    shot options
 *   - clipRect {Object}      截取网页的矩形区域
 *     - top
 *     - left
 *     - width                可以为'window'或者'all',此时可以截取全屏或者全网页
 *     - height               可以为'window'或者'all',此时可以截取全屏或者全网页
 *   - viewportSize {Object}  渲染网页时的分辨率
 *     - width                default 1024
 *     - height               default 768
 *   - renderDelay  {Number}  网页加载完成之后延迟多少毫秒之后截图，默认为0
 *   - mimeType     {String}  生成截图的格式
 *   - script       {Function}网页加载完成之后可以在网页中执行这个方法。
 * @param {Function} callback(err, data)
 *   - data {String} pictrue path
 */
Camera.prototype.shotQN = function (url, qnOptions, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }

  if (!this.qnClient) {
    return callback(new Error('QN not inited'));
  }

  options = options || {};
  qnOptions = qnOptions || {};
  options._type = 'qn';
  var self = this;
  self._handleShot(url, options, function (err, s) {
    if (err) {
      return callback(err);
    }
    console.log(s);
    self.qnClient.upload(s, qnOptions, callback);
  });
};

/**
 * shot web page, genarate a picture Stream.
 * @param {String} url        web page url
 * @param {Object} options    shot options
 *   - clipRect {Object}      截取网页的矩形区域
 *     - top
 *     - left
 *     - width                可以为'window'或者'all',此时可以截取全屏或者全网页
 *     - height               可以为'window'或者'all',此时可以截取全屏或者全网页
 *   - viewportSize {Object}  渲染网页时的分辨率
 *     - width                default 1024
 *     - height               default 768
 *   - renderDelay  {Number}  网页加载完成之后延迟多少毫秒之后截图，默认为0
 *   - mimeType     {String}  生成截图的格式
 *   - script       {Function}网页加载完成之后可以在网页中执行这个方法。
 * @param {Function} callback(err, data)
 *   - data {String} pictrue path
 */
Camera.prototype.shotStream = function (url, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = null;
  }
  options = options || {};
  options._type = 'stream';
  this._handleShot(url, options, callback);
};

exports.create = function (options) {
  return new Camera(options);
};
